File: Python/libraries/datatypes-timex-expression/datatypes_timex_expression/timex_range_resolver.py
from datatypes_timex_expression import Timex, Constants, TimexHelpers, TimeRange, Time, DateRange, TimexDateHelpers
from datatypes_timex_expression.timex_constraints_helper import TimexConstraintsHelper
import copy

class TimexRangeResolver:
    def evaluate(self, candidates: [str], constraints: [str]):
        timex_constraints = list(map(lambda tc: Timex(tc), constraints))
        candidates_with_duration_resolved = self.resolve_duration(
            candidates, timex_constraints)
        candidates_according_to_date = self.resolve_by_date_range_constraints(
            (candidates_with_duration_resolved, timex_constraints))
        candidates_with_added_time = self.resolve_by_time_constraints(
            candidates_according_to_date, timex_constraints)
        candidates_filtered_by_time = self.resolve_by_date_range_constraints(
            candidates_with_added_time, timex_constraints)

        timex_results = list(
            map(lambda tc: Timex(tc), candidates_filtered_by_time))

        return timex_results

    def resolve_durations(self, candidates: [str], constraints: [Timex]):
        results = []
        for candidate in candidates:
            timex = Timex(candidate)

            if Constants.TIMEX_TYPES_DURATION in timex.types():
                resolved_duration = self.resolve_duration(timex, constraints)

                for resolved in resolved_duration:
                    results.append(resolved.timex_value())
            else:
                results.append(candidate)
        return results

    @staticmethod
    def resolve_duration(candidate: Timex, constraints: [Timex]):
        results = []

        for constraint in constraints:
            if Constants.TIMEX_TYPES_DATETIME in constraint.types():
                results.append(TimexHelpers.timex_datetime_add(
                    constraint, candidate))
            elif Constants.TIMEX_TYPES_TIME in constraint.types():
                results.append(TimexHelpers.timex_time_add(
                    constraint, candidate))

        return results

    def resolve_by_date_range_constraints(self, candidates: [str], timex_constraints: [Timex]):
        date_range_constraints = map(lambda t: TimexHelpers.daterange_from_timex(t),
                                     filter(lambda t: Constants.TIMEX_TYPES_DATERANGE in t.types(), timex_constraints))

        collapsed_date_ranges = TimexConstraintsHelper.collapse(TimexConstraintsHelper(), date_range_constraints)

        if not any(collapsed_date_ranges):
            return candidates

        resolution = []

        for timex in candidates:
            resolve_date = self.resolve_date(
                Timex(timex), collapsed_date_ranges)
            resolution.extend(resolve_date)

        return self.remove_duplicates(resolution)

    def resolve_date(self, timex: Timex, constraints):
        result = []

        for constraint in constraints:
            result.extend(
                self.resolve_date_against_constraint(timex, constraint))

        return result

    def resolve_by_timerange_constraints(self, candidates, timex_constraints):
        timerange_contraints = list(map(lambda ti: TimexHelpers.timerange_from_timex(ti),
                                        filter(lambda t: Constants.TIMEX_TYPES_TIMERANGE in t.types(),timex_constraints)))
        collapsed_time_ranges = TimexConstraintsHelper.collapse(TimexConstraintsHelper(),  timerange_contraints)

        if not any(collapsed_time_ranges):
            return candidates

        resolution = []

        for timex in candidates:
            t = Timex(timex)

            if Constants.TIMEX_TYPES_TIMERANGE in t.types():
                r = self.resolve_timerage(t, collapsed_time_ranges)
                resolution.extend(r)
            elif Constants.TIMEX_TYPES_TIME in t.types():
                r = self.resolve_time(t, collapsed_time_ranges)
                resolution.extend(r)

        return self.remove_duplicates(resolution)

    @staticmethod
    def resolve_timerage(timex: Timex, constraints: [TimeRange]):
        candidate = TimexHelpers.timerange_from_timex(timex)
        result = []

        for contraint in constraints:
            if TimexConstraintsHelper.is_overlapping(candidate, contraint):
                start = max(candidate.start.get_time(), contraint.start.get_time())
                time = Time.from_seconds(start)

                resolved = copy.copy(timex)

                resolved.part_of_day = None
                resolved.seconds = None
                resolved.minutes = None
                resolved.hours = None
                resolved.second = time.second
                resolved.minute = time.minute
                resolved.hour = time.hour

                result.append(resolved.timex_value())

        return result

    def resolve_time(self, timex: Timex, constraints: [TimeRange]):
        result = []

        for constraint in constraints:
            result.extend(self.resolve_time_against_constraint(timex, constraint))

        return result

    @staticmethod
    def remove_duplicates(original):
        return list(dict.fromkeys(original))

    @staticmethod
    def resolve_definite_against_constraint(timex, constraint):
        timex_date = TimexHelpers.date_from_timex(timex)
        if constraint.start <= timex_date < constraint.end:
            return [timex.timex_value()]
        return ['']

    def resolve_date_against_constraint(self, timex: Timex, constraint: DateRange):
        if timex.month is not None and timex.day_of_month is not None:
            result = [str]
            year = constraint.start.year
            for year in range(year, year <= constraint.end.year):
                t = copy.copy(timex)
                t.year = year
                result.extend(self.resolve_definite_against_constraint(t, constraint))

            return result

        if timex.day_of_week is not None:
            day = Constants.DAYS['SUNDAY'] if timex.day_of_week == 7 else Constants.DAYS[timex.day_of_week]
            dates = TimexDateHelpers.dates_matching_day(day, constraint.start, constraint.end)
            result = []

            for date in dates:
                t = copy.copy(timex)
                t.day_of_week = None
                t.year = date.year
                t.month = date.month
                t.day_of_month = date.day
                result.append(t.timex_value())
            return result

        return ['']

    def resolve_by_time_constraints(self, candidates, timex_constraints):
        times = list((map( lambda ti: TimexHelpers.time_from_timex(ti),
                           filter(lambda t: Constants.TIMEX_TYPES_TIME in t.types(), timex_constraints))))
        if not any(times):
            return candidates

        resolutions = []

        for timex in filter(lambda t: Timex(t), candidates):
            if Constants.TIMEX_TYPES_DATE in timex.types() and not Constants.TIMEX_TYPES_TIME in timex.types():
                for time in times:
                    timex.hour = time.hour
                    timex.minute = time.minute
                    timex.second = time.second
                    resolutions.append(timex.timex_value())
            else:
                resolutions.append(timex.timex_value())

        return self.remove_duplicates(resolutions)

    @staticmethod
    def resolve_time_against_constraint(timex, constraint):
        t = Time(timex.hour, timex.minute, timex.second)

        if constraint.start.get_time() <= t.get_time() < constraint.end.get_time():
            return [timex.timex_value()]
        return ['']

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexRangeResolver.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.datatypes.timex.expression;

import java.time.DayOfWeek;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class TimexRangeResolver {
    public static List<TimexProperty> evaluate(List<String> candidates, List<String> constraints) {
        List<TimexProperty> timexConstraints = constraints.stream().map(x -> {
            return new TimexProperty(x);
        }).collect(Collectors.toList());
        List<String> candidatesWithDurationsResolved = TimexRangeResolver.resolveDurations(candidates,
                timexConstraints);
        List<String> candidatesAccordingToDate = TimexRangeResolver
                .resolveByDateRangeConstraints(candidatesWithDurationsResolved, timexConstraints);
        List<String> candidatesWithAddedTime = TimexRangeResolver.resolveByTimeConstraints(candidatesAccordingToDate,
                timexConstraints);
        List<String> candidatesFilteredByTime = TimexRangeResolver
                .resolveByTimerangeConstraitns(candidatesWithAddedTime, timexConstraints);

        List<TimexProperty> timexResults = candidatesFilteredByTime.stream().map(x -> {
            return new TimexProperty(x);
        }).collect(Collectors.toList());

        return timexResults;
    }

    public static List<String> resolveDurations(List<String> candidates, List<TimexProperty> constraints) {
        List<String> results = new ArrayList<String>();
        for (String candidate : candidates) {
            TimexProperty timex = new TimexProperty(candidate);
            if (timex.getTypes().contains(Constants.TimexTypes.DURATION)) {
                List<TimexProperty> r = TimexRangeResolver.resolveDuration(timex, constraints);
                for (TimexProperty resolved : r) {
                    results.add(resolved.getTimexValue());
                }
            } else {
                results.add(candidate);
            }
        }

        return results;
    }

    private static List<TimexProperty> resolveDuration(TimexProperty candidate, List<TimexProperty> constraints) {
        List<TimexProperty> results = new ArrayList<TimexProperty>();
        for (TimexProperty constraint : constraints) {
            if (constraint.getTypes().contains(Constants.TimexTypes.DATE_TIME)) {
                results.add(TimexHelpers.timexDateTimeAdd(constraint, candidate));
            } else if (constraint.getTypes().contains(Constants.TimexTypes.TIME)) {
                results.add(TimexHelpers.timexTimeAdd(constraint, candidate));
            }
        }

        return results;
    }

    private static List<String> resolveByDateRangeConstraints(List<String> candidates,
            List<TimexProperty> timexConstraints) {
        List<DateRange> dateRangeconstraints = timexConstraints.stream().filter(timex -> {
            return timex.getTypes().contains(Constants.TimexTypes.DATE_RANGE);
        }).map(timex -> {
            return TimexHelpers.dateRangeFromTimex(timex);
        }).collect(Collectors.toList());

        List<DateRange> collapseDateRanges = TimexConstraintsHelper.collapse(dateRangeconstraints);

        if (collapseDateRanges.isEmpty()) {
            return candidates;
        }

        List<String> resolution = new ArrayList<String>();
        for (String timex : candidates) {
            List<String> r = TimexRangeResolver.resolveDate(new TimexProperty(timex), collapseDateRanges);
            resolution.addAll(r);
        }

        return TimexRangeResolver.removeDuplicates(resolution);
    }

    private static List<String> resolveDate(TimexProperty timex, List<DateRange> constraints) {
        List<String> result = new ArrayList<String>();
        for (DateRange constraint : constraints) {
            result.addAll(TimexRangeResolver.resolveDateAgainstConstraint(timex, constraint));
        }

        return result;
    }

    private static List<String> resolveByTimeRangeConstraints(List<String> candidates,
            List<TimexProperty> timexConstrainst) {
        List<TimeRange> timeRangeConstraints = timexConstrainst.stream().filter(timex -> {
            return timex.getTypes().contains(Constants.TimexTypes.TIME_RANGE);
        }).map(timex -> {
            return TimexHelpers.timeRangeFromTimex(timex);
        }).collect(Collectors.toList());

        List<TimeRange> collapsedTimeRanges = TimexConstraintsHelper.collapse(timeRangeConstraints);

        if (collapsedTimeRanges.isEmpty()) {
            return candidates;
        }

        List<String> resolution = new ArrayList<String>();
        for (String timex : candidates) {
            TimexProperty t = new TimexProperty(timex);
            if (t.getTypes().contains(Constants.TimexTypes.TIME_RANGE)) {
                List<String> r = TimexRangeResolver.resolveTimeRange(t, collapsedTimeRanges);
                resolution.addAll(r);
            } else if (t.getTypes().contains(Constants.TimexTypes.TIME)) {
                List<String> r = TimexRangeResolver.resolveTime(t, collapsedTimeRanges);
                resolution.addAll(r);
            }
        }

        return TimexRangeResolver.removeDuplicates(resolution);
    }

    private static List<String> resolveTimeRange(TimexProperty timex, List<TimeRange> constraints) {
        TimeRange candidate = TimexHelpers.timeRangeFromTimex(timex);

        List<String> result = new ArrayList<String>();
        for (TimeRange constraint : constraints) {
            if (TimexConstraintsHelper.isOverlapping(candidate, constraint)) {
                Integer start = Math.max(candidate.getStart().getTime(), constraint.getStart().getTime());
                Time time = new Time(start);

                // TODO: consider a method on TimexProperty to do this clone/overwrite pattern
                TimexProperty resolved = timex.clone();
                resolved.setPartOfDay(null);
                resolved.setSeconds(null);
                resolved.setMinutes(null);
                resolved.setHours(null);
                resolved.setSecond(time.getSecond());
                resolved.setMinute(time.getMinute());
                resolved.setHour(time.getHour());

                result.add(resolved.getTimexValue());
            }
        }

        return result;
    }

    private static List<String> resolveTime(TimexProperty timex, List<TimeRange> constraints) {
        List<String> result = new ArrayList<String>();
        for (TimeRange constraint : constraints) {
            result.addAll(TimexRangeResolver.resolveTimeAgainstConstraint(timex, constraint));
        }

        return result;
    }

    private static List<String> resolveTimeAgainstConstraint(TimexProperty timex, TimeRange constraint) {
        Time t = new Time(timex.getHour(), timex.getMinute(), timex.getSecond());
        if (t.getTime() >= constraint.getStart().getTime() && t.getTime() < constraint.getEnd().getTime()) {
            return new ArrayList<String>() {
                {
                    add(timex.getTimexValue());
                }
            };
        }

        return new ArrayList<String>();
    }

    private static Set<String> removeDuplicates(List<String> original) {
        return new HashSet<String>(original);
    }

    private static List<String> resolveDefiniteAgainstConstraint(TimexProperty timex, DateRange constraint) {
        DateObject timexDate = TimexHelpers.dateFromTimex(timex);
        if (timexDate >= constraint.getStart() && timexDate < constraint.getEnd()) {
            return new ArrayList<String>() {
                {
                    add(timex.getTimexValue());
                }
            };
        }

        return new ArrayList<String>();
    }

    private static List<String> resolveDateAgainstConstraint(TimexProperty timex, DateRange constraint) {
        if (timex.getMonth() != null && timex.getDayOfMonth() != null) {
            List<String> result = new ArrayList<String>();
            for (int year = constraint.getStart().getYear(); year < constraint.getEnd().getYear(); year++) {
                TimexProperty t = timex.clone();
                t.setYear(year);
                result.addAll(TimexRangeResolver.resolveDefiniteAgainstConstraint(timex, constraint));
            }

            return result;
        }

        if (timex.getDayOfWeek() != null) {
            // convert between ISO day of week and .NET day of week
            DayOfWeek day = timex.getDayOfWeek() == 7 ? DayOfWeek.SUNDAY : DayOfWeek.of(timex.getDayOfWeek());
            List<DateObject> dates = TimexDateHelpers.datesMatchingDay(day, constraint.getStart(), constraint.getEnd());
            List<String> result = new ArrayList<String>();

            for (DateObject d : dates) {
                TimexProperty t = timex.clone();
                t.setDayOfWeek(null);
                t.setYear(d.getYear());
                t.setMonth(d.getMonth());
                t.setDayOfMonth(d.getDay());
                result.add(t.getTimexValue());
            }

            return result;
        }

        if (timex.getHour() != null) {
            List<String> result = new ArrayList<String>();
            DateTime day = constraint.getStart();
            while (day <= constraint.getEnd()) {
                TimexProperty t = timex.clone();
                t.setYear(day.getYear());
                t.setMonth(day.getMonth());
                t.setDayOfMonth(day.getDayOfMonth());
                result.addAll(TimexRangeResolver.resolveDefiniteAgainstConstraint(t, constraint));
                day = day.addDays(1);
            }

            return result;
        }

        return new ArrayList<String>();
    }

    private static List<String> resolveByTimeConstraints(List<String> candidates,
            List<TimexProperty> timexConstrainst) {
        List<Time> times = timexConstrainst.stream().filter(timex -> {
            return timex.getTypes().contains(Constants.TimexTypes.TIME);
        }).map(timex -> {
            return TimexHelpers.timeFromTimex(timex);
        }).collect(Collectors.toList());

        if (times.isEmpty()) {
            return candidates;
        }

        List<String> resolution = new ArrayList<String>();
        for (TimexProperty timex : candidates.stream().map(t -> new TimexProperty(t)).collect(Collectors.toList())) {
            if (timex.getTypes().contains(Constants.TimexTypes.DATE) && !timex.getTypes().contains(Constants.TimexTypes.TIME))) {
                for (Time time : times) {
                    timex.setHour(time.getHour());
                    timex.setMinute(time.getMinute());
                    timex.setSecond(time.getSecond());
                    resolution.add(timex.getTimexValue());
                }
            } else {
                resolution.add(timex.getTimexValue());
            }
        }

        return TimexRangeResolver.removeDuplicates(resolution);
    }
}
